src/python/pants/engine/target.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from abc import ABC, ABCMeta, abstractmethod
+from dataclasses import dataclass
+from typing import Any, ClassVar, Dict, Iterable, Optional, Tuple, Type, TypeVar, cast
+
+from pants.engine.rules import UnionMembership
+from pants.util.memo import memoized_property
+from pants.util.meta import frozen_after_init
+
+
+@dataclass(frozen=True)
+class Field(ABC):
+    alias: ClassVar[str]
+    raw_value: Optional[Any]  # None indicates that the field was not explicitly defined
+
+    def __repr__(self) -> str:
+        return f"{self.__class__}(alias={repr(self.alias)}, raw_value={self.raw_value})"
+
+
+class PrimitiveField(Field, metaclass=ABCMeta):
+    """A Field that does not need the engine in order to be hydrated.
+
+    This should be subclassed by the majority of fields.
+    """
+
+    def __str__(self) -> str:
+        return f"{self.alias}={self.value}"
+
+    @memoized_property
+    @abstractmethod
+    def value(self) -> Any:
+        """Get the field's value.
+
+        The value will possibly be first hydrated and/or validated, such as using a default value
+        if the field was not defined or ensuring that an int value is positive.
+
+        This property is memoized because hydration and validation can often be costly. This
```

Comment:
This is interesting... v2 rules are a mechanism we've built for caching things which may be costly, and which we already _need_ to use for other field types.

What's the argument for not just making all of this caching happen in the v2 graph, and having people access all fields the same, whether or not the engine is just caching or also doing other heavy lifting? Just that the `await Get` feels heavy to have to write?

Reply:
Not only is the `await Get` heavy, but adding a rule for every single `Field` is extremely heavy-weight, both with boilerplate and with blowing up the rule graph. For example, we'd need to introduce a `Result` class for each of the `PrimitiveField`s now so that the graph has a product - we couldn't use a generic wrapper like `FieldResult` because then MyPy couldn't infer the underlying type, and 100% MyPy support is a crucial requirement of this project. 

Compare:

```python
compatibility: List[str] = my_tgt.get(Compatibility).value
```

vs.

```python
compatibility_result = await Get[CompatibilityResult](Compatibility, my_tgt.get(Compatiblity)
compatibility: List[str] = compatibility_result.value
```

Also, it's worth noting that the majority of the hydration/validation for `PrimitiveField` will be exceptionally simple. E.g. `if the field was not defined, use this default value.` Many fields will simply return `self.raw_value`